include/oneapi/dpl/internal/random_impl/philox_engine.h

```diff
@@ -0,0 +1,544 @@
+    {
+        discard_internal(__z);
+    }
+
+    /* Equality operators */
+    friend bool
+    operator==(const philox_engine& __x, const philox_engine& __y)
+    {
+        return (std::equal(__x.state_.X.begin(), __x.state_.X.end(), __y.state_.X.begin()) &&
+                std::equal(__x.state_.K.begin(), __x.state_.K.end(), __y.state_.K.begin()) &&
+                std::equal(__x.state_.Y.begin(), __x.state_.Y.end(), __y.state_.Y.begin()) &&
+                __x.state_.idx == __y.state_.idx);
+    }
+    friend bool
+    operator!=(const philox_engine& __x, const philox_engine& __y)
+    {
+        return !(__x == __y);
+    }
+
+    /* Inserters and extractors */
+    template <typename __CharT, typename __Traits, typename __UIntType, std::size_t __w, std::size_t __n, std::size_t __r,
+              internal::element_type_t<__UIntType>... __consts>
+    friend std::basic_ostream<__CharT, __Traits>&
+    operator<<(std::basic_ostream<__CharT, __Traits>&, const philox_engine<__UIntType, __w, __n, __r, __consts...>&);
+
+    template <typename __UIntType, std::size_t __w, std::size_t __n, std::size_t __r, internal::element_type_t<__UIntType>... __consts>
+    friend const sycl::stream&
+    operator<<(const sycl::stream&, const philox_engine<__UIntType, __w, __n, __r, __consts...>&);
+
+    template <typename __CharT, typename __Traits, typename __UIntType, std::size_t __w, std::size_t __n, std::size_t __r,
+              internal::element_type_t<__UIntType>... __consts>
+    friend std::basic_istream<__CharT, __Traits>&
+    operator>>(std::basic_istream<__CharT, __Traits>&, philox_engine<__UIntType, __w, __n, __r, __consts...>&);
+
+  private:
+    /* Internal generator state */
+    struct state
+    {
+        std::array<scalar_type, word_count> X;     // counters
+        std::array<scalar_type, word_count / 2> K; // keys
+        std::array<scalar_type, word_count> Y;     // results
+        scalar_type idx;                           // index
+    } state_;
+
+    /* __word_mask<_W> - scalar_type with the low _W bits set */
+    template <std::size_t _W>
+    static constexpr scalar_type __word_mask = _W ? (~scalar_type(0) >> (std::numeric_limits<scalar_type>::digits - _W)) : 0;
+
+    /* Processing mask */
+    static constexpr auto in_mask = __word_mask<word_size>;
+
+    void
+    seed_internal(scalar_type __seed)
+    {
+        // set to zero counters and results
+        for (std::size_t __i = 0; __i < word_count; __i++)
+        {
+            state_.X[__i] = 0;
+            state_.Y[__i] = 0;
+        }
+        // 0th key element is set as seed, others are 0
+        state_.K[0] = __seed & in_mask;
+        for (std::size_t __i = 1; __i < (word_count / 2); __i++)
+        {
+            state_.K[__i] = 0;
+        }
+
+        state_.idx = word_count;
+    }
+
+    /* Increment counter by 1 */
+    void
+    increase_counter_internal()
+    {
+        state_.X[0] = (state_.X[0] + 1) & in_mask;
```

Comment:
minor: can be reordered just to not duplicate the same line

```
for (std::size_t __i = 0; __i < word_count; ++__i)
 {
          state_.X[__i] = (state_.X[__i] + 1) & in_mask;
          if (state_.X[__i])
          {
              return;
          }
}
 ```

Reply:
Thanks for the suggestion, applied!